packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -154,9 +149,23 @@ contract Bridge is EssentialContract, IBridge {
         bool support =
             message.from.supportsInterface(type(IRecallableSender).interfaceId);
         if (support) {
+            _ctx = Context({
+                msgHash: msgHash,
+                from: address(this),
+                srcChainId: message.srcChainId
+            });
```

Comment:
I guess for `onMessageRecalled` the context could simply be passed into the function because of the fixed function call, but makes sense to use the same mechanism as the normal call.

Reply:
Not sure I understand you correctly. I think vaults must call back to the Bridge through `context()` to make sure the onMessageCalled is not only called by bridge, but only called within the `recallMessage` function.

Reply:
Isn't that special case handled by having the `assert(message.from != address(this));` in the normal `_invokeMessageCall`? So the context from the bridge can only contain the bridge address itself for `recallMessage`?

Not suggesting any changes, but still interesting to know to make sure I fully understand things. :)

Reply:
Oh never mind, I see know why that can't work!

Reply:
In `sendMessage`, we have `_message.from = msg.sender;` this means message.from should never be "address(this)" -- so the assert will never be triggered.

 But a user can still specify `message.to` to a vault, and the `message.data= abi.encode(Vault.onMessageRecalled.selector, ...)`.  Since the vault trusts the Bridge -- its `recallMessage` function has a `onlyFromNamed("bridge")` modifier, the `onMessageRecalled` can be executed if the user process the message on the destination.

So any bridge-enabled app, all the current token vaults included, must verify that certain functions are not only callable by the bridge, but only by the bridge's processMessage or recallMessage function, not both.

Reply:
The security is so important, do you think we should set up a call to discuss about it?

Reply:
Also, I realized some design assmptions are not valid for enabling the sharing of certain components, such as EtherVault, token vaults, signal service, the bridge, among multiple L2s. To truly make L2->L2 transfer of asset possible, there may be quite some changes... Still working on it.

Reply:
> The security is so important, do you think we should set up a call to discuss about it?

Looking at these issues, I realized I'm really not familiar with how the bridge works, so for me I think it would be best to first spend some more time learning the latest code.

If there's more changes coming, maybe we can do a call after everybody had time to look into those? Or we need a bridge session in Turkey. :)

Reply:
Ok, there will be more changes, mostly to:

1) simplify the vault codebase
2) enable L2-L2 bridging (still not working)

Reply:
Let me merge this one first.